From d460ce7ce0e19a42cf30d9b24b7f2d1abc4685a8 Mon Sep 17 00:00:00 2001
From: Ioana Radulescu <ruxandra.radulescu@nxp.com>
Date: Thu, 4 May 2017 15:56:01 +0300
Subject: [PATCH 140/448] staging: fsl-dpaa2/eth: Read link state in ethtool get_settings

dpni_set_link_cfg() needs to be called with all options,
including the ones involving flow control. Without this
patch, we risk inadvertently disabling pause frames while
configuring other link parameters.

Signed-off-by: Bogdan Purcareata <bogdan.purcareata@nxp.com>
Signed-off-by: Ioana Radulescu <ruxandra.radulescu@nxp.com>
---
 drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c |   12 +++++++++++-
 1 files changed, 11 insertions(+), 1 deletions(-)

diff --git a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c
index 0fca26a..f6388cb 100644
--- a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c
+++ b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-ethtool.c
@@ -118,8 +118,9 @@ static void dpaa2_eth_get_drvinfo(struct net_device *net_dev,
 dpaa2_eth_set_link_ksettings(struct net_device *net_dev,
 			     const struct ethtool_link_ksettings *link_settings)
 {
-	struct dpni_link_cfg cfg = {0};
 	struct dpaa2_eth_priv *priv = netdev_priv(net_dev);
+	struct dpni_link_state state = {0};
+	struct dpni_link_cfg cfg = {0};
 	int err = 0;
 
 	netdev_dbg(net_dev, "Setting link parameters...");
@@ -133,6 +134,14 @@ static void dpaa2_eth_get_drvinfo(struct net_device *net_dev,
 		return -EACCES;
 	}
 
+	/* Need to interrogate link state to get flow control params */
+	err = dpni_get_link_state(priv->mc_io, 0, priv->mc_token, &state);
+	if (err) {
+		netdev_err(net_dev, "Error getting link state\n");
+		goto out;
+	}
+
+	cfg.options = state.options;
 	cfg.rate = link_settings->base.speed;
 	if (link_settings->base.autoneg == AUTONEG_ENABLE)
 		cfg.options |= DPNI_LINK_OPT_AUTONEG;
@@ -150,6 +159,7 @@ static void dpaa2_eth_get_drvinfo(struct net_device *net_dev,
 		 */
 		netdev_dbg(net_dev, "ERROR %d setting link cfg\n", err);
 
+out:
 	return err;
 }
 
-- 
1.7.1

